server/src/main/java/org/elasticsearch/index/mapper/JsonFieldMapper.java

```diff
@@ -334,6 +345,98 @@ public BytesRef indexedValueForSearch(Object value) {
             String keyedValue = JsonFieldParser.createKeyedValue(key, stringValue);
             return new BytesRef(keyedValue);
         }
+
+        @Override
+        public IndexFieldData.Builder fielddataBuilder(String fullyQualifiedIndexName) {
+            failIfNoDocValues();
+            return new KeyedJsonIndexFieldData.Builder(key);
+        }
+    }
+
+    public static class KeyedJsonIndexFieldData implements IndexOrdinalsFieldData {
+        private final String key;
+        private final IndexOrdinalsFieldData delegate;
+
+        private KeyedJsonIndexFieldData(String key, IndexOrdinalsFieldData delegate) {
+            this.delegate = delegate;
+            this.key = key;
+        }
+
+        public String getKey() {
+            return key;
+        }
+
+        @Override
+        public String getFieldName() {
+            return delegate.getFieldName();
+        }
+
+        @Override
+        public SortField sortField(Object missingValue,
+                                   MultiValueMode sortMode,
+                                   XFieldComparatorSource.Nested nested,
+                                   boolean reverse) {
+            return delegate.sortField(missingValue, sortMode, nested, reverse);
```

Comment:
The `SortField` will use the original doc values to perform the sort so you'll need to create a custom `SortedSetSortField` that uses a `KeyedJsonDocValues` in order to filter the values that should not participate in the sort.

Reply:
👍

Reply:
I took a closer look at `SortedSetDVOrdinalsIndexFieldData` to understand how this should be implemented, and was trying to figure out whether this optimization was relevant: https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/index/fielddata/plain/SortedSetDVOrdinalsIndexFieldData.java#L76-L79. Would you be able to give a little bit of background on this logic (originally introduced in #23827)?

Reply:
This is not an optimization but a change that was needed to handle index sorting. Index sorting at the index writer level accepts the Lucene's FieldSort only and since we compare index sort with the query sort to activate early termination we needed to make the query sort compatible. This shouldn't be an issue for the `json` field since this should not be allowed to use it for index sorting.

Reply:
That makes sense, thanks!